// tm-text/dist/cjs/utils/syntax.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SYNTAX_MAP = exports.SYNTAX_MANIAPLANET = exports.SYNTAX_FOREVER = exports.SYNTAX_UNITED = exports.SYNTAX_CLASSIC = exports.SYNTAX = exports.TOKEN_TO_CHAR_MAP = exports.TOKEN = void 0;
exports.TOKEN = {
    BLOCK_END: 'BLOCK_END',
    BLOCK_START: 'BLOCK_START',
    BOLD: 'BOLD',
    COLOR: 'COLOR',
    HREF_CONTENT: 'HREF_CONTENT',
    HREF_START: 'HREF_START',
    HREF_END: 'HREF_END',
    ITALIC: 'ITALIC',
    LINK_EXTERNAL: 'LINK_EXTERNAL',
    LINK_INTERNAL: 'LINK_INTERNAL',
    LINK_INTERNAL_WITH_PARAMS: 'LINK_INTERNAL_WITH_PARAMS',
    NEWLINE: 'NEWLINE',
    RESET_ALL: 'RESET_ALL',
    RESET_COLOR: 'RESET_COLOR',
    SHADOW: 'SHADOW',
    TAB: 'TAB',
    UPPERCASE: 'UPPERCASE',
    WIDTH_NARROW: 'WIDTH_NARROW',
    WIDTH_NORMAL: 'WIDTH_NORMAL',
    WIDTH_WIDE: 'WIDTH_WIDE',
    WORD: 'WORD',
};
exports.TOKEN_TO_CHAR_MAP = {
    [exports.TOKEN.BLOCK_END]: '>',
    [exports.TOKEN.BLOCK_START]: '<',
    [exports.TOKEN.BOLD]: 'o',
    [exports.TOKEN.HREF_START]: '[',
    [exports.TOKEN.HREF_END]: ']',
    [exports.TOKEN.ITALIC]: 'i',
    [exports.TOKEN.LINK_EXTERNAL]: 'l',
    [exports.TOKEN.LINK_INTERNAL]: 'h',
    [exports.TOKEN.LINK_INTERNAL_WITH_PARAMS]: 'p',
    [exports.TOKEN.RESET_ALL]: 'z',
    [exports.TOKEN.RESET_COLOR]: 'g',
    [exports.TOKEN.SHADOW]: 's',
    [exports.TOKEN.UPPERCASE]: 't',
    [exports.TOKEN.WIDTH_NARROW]: 'n',
    [exports.TOKEN.WIDTH_NORMAL]: 'm',
    [exports.TOKEN.WIDTH_WIDE]: 'w',
};
exports.SYNTAX = {
    /**
     * Games that use this syntax: `Original`, `Sunrise`, `Nations`
     */
    CLASSIC: 'CLASSIC',
    /**
     * Games that use this syntax: `United`
     */
    UNITED: 'UNITED',
    /**
     * Games that use this syntax: `United Forever`, `Nations Forever`
     */
    FOREVER: 'FOREVER',
    /**
     * Games that use this syntax: `Maniapanet`, `Turbo`, `2020`
     */
    MANIAPLANET: 'MANIAPLANET',
};
exports.SYNTAX_CLASSIC = [
    exports.TOKEN.COLOR,
    exports.TOKEN.ITALIC,
    exports.TOKEN.RESET_ALL,
    exports.TOKEN.RESET_COLOR,
    exports.TOKEN.NEWLINE,
    exports.TOKEN.SHADOW,
    exports.TOKEN.TAB,
    exports.TOKEN.UPPERCASE,
    exports.TOKEN.WIDTH_NARROW,
    exports.TOKEN.WIDTH_NORMAL,
    exports.TOKEN.WIDTH_WIDE,
    exports.TOKEN.WORD,
];
exports.SYNTAX_UNITED = [
    ...exports.SYNTAX_CLASSIC,
    exports.TOKEN.BOLD,
    exports.TOKEN.HREF_CONTENT,
    exports.TOKEN.HREF_START,
    exports.TOKEN.HREF_END,
    exports.TOKEN.LINK_EXTERNAL,
    exports.TOKEN.LINK_INTERNAL,
];
exports.SYNTAX_FOREVER = [
    ...exports.SYNTAX_UNITED,
    exports.TOKEN.LINK_INTERNAL_WITH_PARAMS,
];
exports.SYNTAX_MANIAPLANET = [
    ...exports.SYNTAX_FOREVER,
    exports.TOKEN.BLOCK_END,
    exports.TOKEN.BLOCK_START,
];
exports.SYNTAX_MAP = {
    [exports.SYNTAX.CLASSIC]: exports.SYNTAX_CLASSIC,
    [exports.SYNTAX.UNITED]: exports.SYNTAX_UNITED,
    [exports.SYNTAX.FOREVER]: exports.SYNTAX_FOREVER,
    [exports.SYNTAX.MANIAPLANET]: exports.SYNTAX_MANIAPLANET,
};
